src/visitors/sympy_replace_solutions_visitor.cpp

```diff
@@ -0,0 +1,350 @@
+    interleaves_counter_.new_equation(true);
+
+    const auto& statement = std::static_pointer_cast<ast::Statement>(
+        node.get_parent()->get_shared_ptr());
+
+    // do not visit if already marked
+    if (replacements_.find(statement) != replacements_.end()) {
+        return;
+    }
+
+    switch (policy_) {
+    case ReplacePolicy::VALUE: {
+        const auto dependencies = statement_dependencies(get_lhs(node), get_rhs(node));
+        const auto& key = dependencies.first;
+        const auto& vars = dependencies.second;
+
+        if (solutions_.is_var_assigned_here(key)) {
+            ast::StatementVector new_statements;
+
+            pre_solve_statements_.emplace_back_all_tagged_statements(new_statements);
+            tmp_statements_.emplace_back_all_tagged_statements(new_statements);
+            solutions_.try_emplace_back_tagged_statement(new_statements, key);
+
+            replacements_.emplace(statement, new_statements);
+        }
+        break;
+    }
+    case ReplacePolicy::GREEDY: {
+        ast::StatementVector new_statements;
+
+        pre_solve_statements_.emplace_back_all_tagged_statements(new_statements);
+        tmp_statements_.emplace_back_all_tagged_statements(new_statements);
+        solutions_.emplace_back_next_tagged_statements(new_statements, n_next_equations_);
+
+        replacements_.emplace(statement, new_statements);
+
+        break;
+    }
+    }
+}
+
+
+void SympyReplaceSolutionsVisitor::visit_diff_eq_expression(ast::DiffEqExpression& node) {
+    auto get_lhs = [](const ast::Node& node) -> const std::shared_ptr<ast::Expression>& {
+        return static_cast<const ast::DiffEqExpression&>(node).get_expression()->get_lhs();
+    };
+
+    auto get_rhs = [](const ast::Node& node) -> const std::shared_ptr<ast::Expression>& {
+        return static_cast<const ast::DiffEqExpression&>(node).get_expression()->get_rhs();
+    };
+
+    try_replace_tagged_statement(node, get_lhs, get_rhs);
+}
+
+void SympyReplaceSolutionsVisitor::visit_lin_equation(ast::LinEquation& node) {
+    auto get_lhs = [](const ast::Node& node) -> const std::shared_ptr<ast::Expression>& {
+        return static_cast<const ast::LinEquation&>(node).get_left_linxpression();
+    };
+
+    auto get_rhs = [](const ast::Node& node) -> const std::shared_ptr<ast::Expression>& {
+        return static_cast<const ast::LinEquation&>(node).get_left_linxpression();
+    };
+
+    try_replace_tagged_statement(node, get_lhs, get_rhs);
+}
+
+
+void SympyReplaceSolutionsVisitor::visit_non_lin_equation(ast::NonLinEquation& node) {
+    auto get_lhs = [](const ast::Node& node) -> const std::shared_ptr<ast::Expression>& {
+        return static_cast<const ast::NonLinEquation&>(node).get_lhs();
+    };
+
+    auto get_rhs = [](const ast::Node& node) -> const std::shared_ptr<ast::Expression>& {
+        return static_cast<const ast::NonLinEquation&>(node).get_rhs();
+    };
+
+    try_replace_tagged_statement(node, get_lhs, get_rhs);
+}
```

Comment:
it is less verbose but I call `statement_dependencies` even if the greedy approach is employed. in this way it is more efficent because I call it only when I need it. What do you prefer? (I do not have strong opinions on this)